### ISSUE TITLE
elm-test hanging on circleci

### ISSUE THREAD

**Issue by 97jaz** [rtfeldman/node-test-runner#201]
(And yes, I'm running it with `sysconfcpus -n 2 elm-test`.)

I only have two tests at the moment. (Unfortunately, the code is not open source, or else I'd post them.)

The closest issue I found was #53, but I'm not running on Windows, and neither of the tests are fuzz tests. The output is:
```
$ #!/bin/bash -eo pipefail
sysconfcpus -n 2 elm-test

Success! Compiled 105 modules.
Successfully generated /dev/null
Success! Compiled 11 modules.
Successfully generated /home/circleci/connect/elm-stuff/generated-code/elm-community/elm-test/elmTestOutput.js

elm-test 0.18.8
---------------

Running 2 tests. To reproduce these results, run: elm-test --fuzz 100 --seed 745527669
```

And then it never completes. Circle CI allows you to ssh into the box to debug a build. When I did that and ran `elm-test` myself, it worked the first time and hung thereafter.

Any thoughts on how to debug this further?

**Comment by 97jaz** [comment]
Oh, and this is probably obvious, but it never hangs when I run it locally (on OS X). The CircleCI container is Ubuntu Trusty. The tests themselves, and the code they test, are deterministic. They should always terminate.

**Comment by 97jaz** [comment]
Okay, it appears that when this happens there are a ton of node processes running, which suggests that the effect of `sysconfcpus -n 2` isn't carried over to whatever subprocesses are created by `elm-test`.

... except I'm pretty sure it's supposed to be. I also tried:
```bash
export LIBSYSCONFCPUS=2
sysconfcpus  elm-test
```
which didn't help.

**Comment by 97jaz** [comment]
And... that's not surprising, since node's `os.cpus()` isn't affected by `sysconfcpus`.

So that's the issue: `sysconfcpus` just doesn't have any effect on how many processors `node` thinks there are: https://github.com/rtfeldman/node-test-runner/blob/master/lib/elm-test.js#L59

Maybe a `--cpus` or `--workers` flag is in order.

(That said, I don't know for a fact that this is what's causing `elm-test` to hang.)

**Comment by 97jaz** [comment]
I apologize for the many comments in succession, but I just tried with version 0.18.7, and it worked, so I do suspect that `os.cpus()` is the culprit here.

**Comment by rtfeldman** [comment]
Hm, looks like Node folks looked into distinguishing between physical and virtual cores, and concluded they aren't going to address it: https://github.com/nodejs/node/issues/7730

**Comment by rtfeldman** [comment]
> And then it never completes. Circle CI allows you to ssh into the box to debug a build. When I did that and ran elm-test myself, it worked the first time and hung thereafter.

This is strange behavior, and it suggests something is getting left over in between runs that affects whether it hangs.

What happens if you delete `elm-stuff` in between runs? (Both `./elm-stuff` and `./tests/elm-stuff` that is.)

Also curious if any processes (`node` or otherwise) are running after the first test run which weren't running before the run started. If Node processes aren't getting cleaned up properly, that could also be a potential culprit.

I'm generally opposed to creating new CLI flags as a bugfixing tactic, so I want to push very hard to try and figure out why this is happening and fix it properly!

**Comment by 97jaz** [comment]
There aren't any multiple runs (except in this atypical case of me ssh-ing into a container), since these test runs are all in freshly-constructed virtual containers. And this problem occurs even when I do not ssh in.

The containers each report having (IIRC) 36 cpus even though only two are actually available to the container. It's the exact same thing that prompted people to start using sysconfcpus to begin with, except that os.cpus doesn't use the system calls that sysconfcpus intercepts.

Also, this isn't an issue of distinguishing between physical and virtual processors but one of determining how many virtual processors are actually available for use. (Ultimately it's a matter of incomplete virtualization.)

**Comment by 97jaz** [comment]
Okay, so here's the output of node's `os.cpus()` on a CircleCI test container:
```js
[ { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1304,
    times: { user: 5790200, nice: 1100, sys: 2340200, idle: 32297500, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 3209,
    times: { user: 4364400, nice: 600, sys: 1435600, idle: 38806800, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 3200,
    times: { user: 4241600, nice: 600, sys: 1311100, idle: 39605700, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 3210,
    times: { user: 3819000, nice: 1000, sys: 1024300, idle: 40590400, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 3200,
    times: { user: 3912000, nice: 100, sys: 1089200, idle: 40518600, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1254,
    times: { user: 4047300, nice: 600, sys: 988800, idle: 40665700, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1485,
    times: { user: 3843900, nice: 300, sys: 1031100, idle: 40587800, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1271,
    times: { user: 3972800, nice: 600, sys: 1065800, idle: 40512900, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 2522,
    times: { user: 3812100, nice: 400, sys: 1108000, idle: 40934500, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1896,
    times: { user: 6026500, nice: 100, sys: 1349200, idle: 38446200, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1896,
    times: { user: 4306800, nice: 500, sys: 1012500, idle: 40443200, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1397,
    times: { user: 4000700, nice: 100, sys: 864800, idle: 40994400, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1476,
    times: { user: 4108200, nice: 200, sys: 936400, idle: 40856500, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1899,
    times: { user: 3655100, nice: 300, sys: 798300, idle: 41493500, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1391,
    times: { user: 3932500, nice: 100, sys: 859000, idle: 41127600, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1493,
    times: { user: 3510400, nice: 300, sys: 783400, idle: 41615800, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1499,
    times: { user: 3541800, nice: 400, sys: 838100, idle: 41609600, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1392,
    times: { user: 3321200, nice: 300, sys: 721800, idle: 41930500, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 3035,
    times: { user: 963700, nice: 500, sys: 299000, idle: 45137300, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1853,
    times: { user: 1160300, nice: 100, sys: 334100, idle: 44923700, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 3194,
    times: { user: 1228200, nice: 300, sys: 368000, idle: 44826700, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 3203,
    times: { user: 1374000, nice: 100, sys: 377300, idle: 44680000, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 3200,
    times: { user: 1356200, nice: 200, sys: 334800, idle: 44747800, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1929,
    times: { user: 1140100, nice: 100, sys: 289200, idle: 45017300, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1697,
    times: { user: 1556200, nice: 100, sys: 289100, idle: 44597700, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1430,
    times: { user: 1160200, nice: 200, sys: 269200, idle: 45033800, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 2578,
    times: { user: 1473100, nice: 400, sys: 446100, idle: 44495100, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1374,
    times: { user: 1030800, nice: 200, sys: 317800, idle: 45132300, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1611,
    times: { user: 997400, nice: 100, sys: 307300, idle: 45205600, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1387,
    times: { user: 1166000, nice: 200, sys: 328000, idle: 44985100, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1387,
    times: { user: 1096900, nice: 200, sys: 324200, idle: 45072000, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1601,
    times: { user: 1024800, nice: 0, sys: 270600, idle: 45220400, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1199,
    times: { user: 1033800, nice: 300, sys: 300600, idle: 45170600, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1273,
    times: { user: 1039400, nice: 0, sys: 297900, idle: 45162400, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1600,
    times: { user: 1064900, nice: 200, sys: 290300, idle: 45148300, irq: 0 } },
  { model: 'Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz',
    speed: 1287,
    times: { user: 1104600, nice: 0, sys: 287600, idle: 45102100, irq: 0 } } ]
```
(And, yeah, there are 36 logical processors reported, even though CircleCI makes 2 available to each container).

When the tests are hung, the process list has 36 lines like this:
```
521 pts/5    Sl+    0:00 /usr/local/bin/node /home/circleci/connect/elm-stuff/generated-code/elm-community/elm-test/elmTestOutput.js
```

Also, they don't always hang, just most of the time (like 80-90% of the time).

Okay, when I delete the `elm-stuff` directories and then run `elm-test`...

Attempt 1:
```
circleci@57930f79ac11:~/connect$ rm -rf elm-stuff/
circleci@57930f79ac11:~/connect$ rm -rf tests/elm-stuff/
circleci@57930f79ac11:~/connect$ sysconfcpus -n 2 elm-test
Success! Compiled 105 modules.                                      
Successfully generated /dev/null
Success! Compiled 11 modules.                                       
Successfully generated /home/circleci/connect/elm-stuff/generated-code/elm-community/elm-test/elmTestOutput.js

elm-test 0.18.8
---------------

Running 2 tests. To reproduce these results, run: elm-test --fuzz 100 --seed 1235356718
^C
```
*tests hang*


Attempt 2 (this time using `sysconfcpus -n 1` in case it makes a difference): *tests hang*
Attempt 3: (not using `sysconfcpus` at all): *compilation is super-slow, tests hang*
Attempt 4: (back to `-n 2`): *hang*
Attempt 5: *hang*

[At this point I stopped deleting `elm-stuff` directories before each run, since that doesn't seem to have any effect.]

6: hang
7: hang
8: *success* -- tests complete:
```
TEST RUN PASSED

Duration: 898 ms
Passed:   2
Failed:   0
```

So there's definitely some kind of non-determinism here. It's still possible that it has nothing to do with Circle's over-reporting of the number of processors, but that should be easy to test with a branch that only creates 1 process.

**Comment by rtfeldman** [comment]
Wow, great to know!

Is it possible for you to edit the code and replace os.cpus() with a
hardcoded 1? (The call only happens in one place in the code base.)

If not, I can publish a version that makes that change.

Thanks for helping diagnose!

**Comment by 97jaz** [comment]
Yeah, thanks for restoring my sanity there, @rtfeldman. I had forked the repo and was about to figure out how to get my test suite to use a local branch. Then I read your message and thought, "Oh, right. I can just edit the code in-place on the test server."

So, I did that, and it never hangs. So, it does appear to be the CPU over-reporting.

Really, the fault is with Circle's incomplete virtualization. If they're only making two processors available to a container, then the system calls that return the processor info should indicate that there are two. But...

So, I'm not sure what the solution is. I don't see a call in node that will be affected by `sysconfcpus`, unfortunately, but I'll do a bit more digging.

**Comment by 97jaz** [comment]
related issue: https://discuss.circleci.com/t/memory-problems-with-jest-and-workers/10297

**Comment by haplesshero13** [comment]
I'm having the same "non-deterministic" hanging problem on a relatively new iMac with macOS Sierra. Using elm-test 0.18.8 and node 8.4.0.

**Comment by 97jaz** [comment]
@haplesshero13 That's interesting, because it suggests that the problem might not be caused specifically by using too many processes. There could just be a concurrency-related bug.

If you're willing, could you try my branch at https://github.com/careport/node-test-runner/tree/processes-flag and run it with `--processes 1` and see if the problem still occurs?

You'd need to use [https://github.com/gdotdesign/elm-github-install] to install the package, since the official Elm package manager doesn't allow you to install non-official, non-released versions of anything. In your `elm-package.json` you'd add a dependency for:
```json
"careport/node-test-runner": {
  "url": "git@github.com:careport/node-test-runner.git",
  "ref": "processes-flag"
}
```

**Comment by crackofdusk** [comment]
@97jaz Here's another data point:

I also get random hangs locally on macOS:

```
$ elm-test
Success! Compiled 1 module.                                         
Successfully generated /dev/null
Success! Compiled 1 module.                                         
Successfully generated $REDACTED/elm-stuff/generated-code/elm-community/elm-test/elmTestOutput.js

elm-test 0.18.8
---------------

Running 1 test. To reproduce these results, run: elm-test --fuzz 100 --seed 2024873990
```

Running https://github.com/careport/node-test-runner/tree/processes-flag with `--process 1` sometimes hangs after the compilation step:

```
$ ./node_modules/.bin/elm-test --processes 1
Success! Compiled 1 module.                                         
Successfully generated /dev/null
Success! Compiled 1 module.                                         
Successfully generated $REDACTED/elm-stuff/generated-code/elm-community/elm-test/elmTestOutput.js
```

It's interesting that they don't hang at the same point.

**Comment by 97jaz** [comment]
@crackofdusk Do the tests hang more often without the `--processes 1` option? I ask because in my case (on CircleCI, in a Linux container), `elm-test` *usually* hangs without when I don't limit the number of processes, and I haven't yet seen it hang when I do (but I also haven't run it enough times to be confident that it *never* does, because I reverted to the previous version of `elm-test` in CI).

**Comment by rtfeldman** [comment]
It would also help to know whether it only ever hangs when 1+ tests fail vs if it sometimes hangs even when all tests pass.

Any info there?

PS: thanks for all the additional data! This is a really tough one to isolate, so I appreciate all the info we can get.

**Comment by crackofdusk** [comment]
It hangs both when all tests pass and when there is a failing test.

**Comment by tonytamps** [comment]
Also experiencing this. Have started to use the `--processes 1` option now and will report back if I notice any hands while using it.

It should raise it's head pretty quickly, we're running the tests many times per day on circle.

**Comment by leifg** [comment]
In case you're using circle 2.0: what image do you use? I have the same problem when using `node:8.4.0` (the official image).

I used `node:8.3.0` and it sometimes worked.

I then switched to `node:8.2.1` and it has been working consistently ever since.

**Comment by 97jaz** [comment]
We're using a custom docker image, but it's based on `circleci/ruby:2.4.1-node`.

**Comment by tonytamps** [comment]
I am using the `circleci/node:8.4.0` image. I haven't had a hang yet using the testrunner from @crackofdusk

**Comment by isaacseymour** [comment]
I think this is the same issue as #219 probably...?

**Comment by rtoal** [comment]
It's probably the same as #219. elm-test hangs the vast majority of times for me, even on an empty project for which I just do
```
$ npm install -g elm-test
$ elm test init
$ elm test
```
and attempt to run the plain old auto generated `Example.elm`.

Eight times I ran elm-test and it hung with

```
$ elm test
Success! Compiled 0 modules.                                        
Successfully generated /dev/null
Success! Compiled 1 module.                                         
Successfully generated /Users/ray/git-repos/lmu-cmsi-386/homework4/elm-stuff/generated-code/elm-community/elm-test/elmTestOutput.js

elm-test 0.18.9
---------------

Running 1 test. To reproduce these results, run: elm-test --fuzz 100 --seed 1772203118
```

(where the seed changed each time of course.) The ninth time I ran it, I FINALLY saw

```
TEST RUN INCOMPLETE because there is 1 TODO remaining

Duration: 182 ms
Passed:   0
Failed:   0
Todo:     1
↓ Example
◦ TODO: Implement our first test. See http://package.elm-lang.org/packages/elm-community/elm-test/latest for how to do this!
```

Definitely looking forward to a fix. In the meantime, my students will not be using elm-test for their Elm assignment. Too bad, because it's really nice.

I am using a MacBookPro running Sierra and Elm 0.18.0.

**Comment by zwilias** [comment]
@rtoal @97jaz can you confirm this is resolved for you with 0.18.10?

**Comment by rtoal** [comment]
Beautiful, thank you, no more hanging!

**Comment by 97jaz** [comment]
@zwilias This only ever failed for me on CircleCI. I'll have to create a new Docker image for Circle to try out 0.18.10, and I don't really have the time for that now. I've long since reverted to 0.18.8, which works fine.

**Comment by zwilias** [comment]
Alright, cool; thanks for the responses! So far the issues seem resolved with 0.18.10, so I'll close this for now.

**Comment by 97jaz** [comment]
@zwilias Just wanted to let you know, I finally got around to upgrading the docker image I was using, and the new version of `node-test-runner` works fine. Thanks again!

**Comment by simonh1000** [comment]
I'm seeing the same now with 0.19 but I'm rather out of my depth here. What I do know is that my code is based on very similar code that has worked well for us at my firm with 0.18. Any ideas?

Dockerfile: https://github.com/simonh1000/elm-webpack-starter/blob/ci/.circleci/Dockerfile
Circleci config: https://github.com/simonh1000/elm-webpack-starter/blob/ci/.circleci/config.yml
package.json
```
        "elm": "^0.19.0-bugfix2",
        "elm-hot-webpack-loader": "^1.0.2",
        "elm-test": "^0.19.0-beta9",
        "elm-verify-examples": "^3.0.1",
        "elm-webpack-loader": "^5.0.0",
```

**Comment by rtfeldman** [comment]
@simonh1000 This was fixed in `elm-test` version `0.19.0-beta10` - can you bump the version and see if it still reproduces?
